src/material-experimental/mdc-dialog/dialog-container.ts

```diff
@@ -0,0 +1,176 @@
+
+    if (this._animationsEnabled) {
+      this._hostElement.classList.add(cssClasses.CLOSING);
+      this._waitForAnimationToComplete(this._closeAnimationDuration, this._finishDialogClose);
+    } else {
+      // This subscription to the `OverlayRef#backdropClick` observable in the `DialogRef` is
+      // set up before any user can subscribe to the backdrop click. The subscription triggers
+      // the dialog close and this method synchronously. If we'd synchronously emit the `CLOSED`
+      // animation state event if animations are disabled, the overlay would be disposed
+      // immediately and all other subscriptions to `DialogRef#backdropClick` would be silently
+      // skipped. We work around this by waiting with the dialog close until the next tick when
+      // all subscriptions have been fired as expected. This is not an ideal solution, but
+      // there doesn't seem to be any other good way. Alternatives that have been considered:
+      //   1. Deferring `DialogRef.close`. This could be a breaking change due to a new microtask.
+      //      Also this issue is specific to the MDC implementation where the dialog could
+      //      technically be closed synchronously. In the non-MDC one, Angular animations are used
+      //      and closing always takes at least a tick.
+      //   2. Ensuring that user subscriptions to `backdropClick`, `keydownEvents` in the dialog
+      //      ref are first. This would solve the issue, but has the risk of memory leaks and also
+      //      doesn't solve the case where consumers call `DialogRef.close` in their subscriptions.
+      // Based on the fact that this is specific to the MDC-based implementation of the dialog
+      // animations, the defer is applied here.
+      Promise.resolve().then(() => this._finishDialogClose());
+    }
+  }
+
+  /**
+   * Completes the dialog open by clearing potential animation classes, trapping
+   * focus and emitting an opened event.
+   */
+  private _finishDialogOpen =
+      () => {
+        this._clearAnimationClasses();
+        this._trapFocus();
+        this._animationStateChanged.emit({state: 'opened', totalTime: this._openAnimationDuration});
+      }
+
+  /**
+   * Completes the dialog close by clearing potential animation classes, restoring
+   * focus and emitting a closed event.
+   */
+  private _finishDialogClose =
+      () => {
+        this._clearAnimationClasses();
+        this._restoreFocus();
+        this._animationStateChanged.emit(
+            {state: 'closed', totalTime: this._closeAnimationDuration});
+      }
+
+  /** Clears all dialog animation classes. */
+  private _clearAnimationClasses() {
+    this._hostElement.classList.remove(cssClasses.OPENING);
+    this._hostElement.classList.remove(cssClasses.CLOSING);
+  }
+
+  private _waitForAnimationToComplete(duration: number, callback: () => void) {
+    this._ngZone.runOutsideAngular(() => this._animationTimer = setTimeout(callback, duration));
```

Comment:
Should we `clearTimeout` here to avoid any leftovers?

Reply:
Sounds good.